### ISSUE TITLE
BadSetOnceAccess on invalid double assignment involving `new`

### ISSUE THREAD

**Issue by agroce** [ethereum/solidity#9952]
## Description

This contract:
```solidity
library Pairing {
    struct G1Point {}
    struct G2Point {}
    function pairingProd4(
	G1Point memory a1,
	G2Point memory d2
    )
    {
	G2Point[] memory p2 = new G2Point[]= a1;
    }
}
```

produces:

```
Exception during compilation: /home/user/solidity/libsolutil/SetOnce.h(73): Throw in function const T &solidity::util::SetOnce<bool>::operator*() const [T = bool]
Dynamic exception type: boost::exception_detail::clone_impl<solidity::util::BadSetOnceAccess>
std::exception::what: BadSetOnceAccess
[solidity::util::tag_comment*] = Attempt to access the value of a SetOnce that does not have a value.
```

when compiled with solc.

Probably related to https://github.com/ethereum/solidity/issues/9890 -- or at least the same symptom.

On master, using AFL fuzzing.  Another discovery using https://github.com/agroce/afl-compiler-fuzzer.

## Environment

- Compiler version:  0.7.3-develop.2020.10.2+commit.756e21a8.Linux.clang
- Target EVM version (as per compiler settings): N/A
- Framework/IDE (e.g. Truffle or Remix): N/A
- EVM execution environment / backend / blockchain client: N/A
- Operating system: Ubuntu 18.04 in docker

## Steps to Reproduce

Above shows pretty clearly, I think.

**Comment by cameel** [comment]
Thanks for the report. Your fuzzing seems pretty effective :)

Yeah, it does look like another corner case from #9839 where a `SetOnce` value gets assigned twice (pinging @Marenz again). The control flow in that PR was pretty hard to analyze (lots of branches in many functions calling each other) so this might not be the last one. It started producing an ICE on 0.7.2 where that PR was merged.

### Simplified sample
The key elements here seem to be the double assignments and the use of `new`:

```solidity
contract C {
    function foo(uint x)
    {
        uint p = new uint[] = x;
    }
}
```

### Errors on 0.7.1 for the simplified sample
```
Error: No visibility specified. Did you intend to add "public"?
 --> /tmp/test.sol:2:5:
  |
2 |     function foo(uint x)
  |     ^ (Relevant source part starts here and spans across multiple lines).

Error: Expression has to be an lvalue.
 --> /tmp/test.sol:4:18:
  |
4 |         uint p = new uint[] = x;
  |                  ^^^^^^^^^^

Error: Type uint256 is not implicitly convertible to expected type function (uint256) pure returns (uint256[] memory).
 --> /tmp/test.sol:4:31:
  |
4 |         uint p = new uint[] = x;
  |                               ^

Error: Type function (uint256) pure returns (uint256[] memory) is not implicitly convertible to expected type uint256.
 --> /tmp/test.sol:4:9:
  |
4 |         uint p = new uint[] = x;
  |         ^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by bshastry** [comment]
~~A smaller repro~~ Some more repros. This crash occurs within seconds and very frequently in our fuzzing rigs.

```
contract A{}contract B{A a=new B=a;}
```

```
enum X{A}function t(){X.A=X;}
```

```
contract A{}contract B{A a=new B=();}
```

```
contract C{function f(){new int[]=6;}}
```

```
contract C{function f(){new int[]=int;}}
```

```
contract C{
  enum A {X}
  function f() {(delete A.X,[]);}
}
```

```
contract D{function foo(){this.foo{g:5}=();}}
```

```
contract C{struct T{ }enum A{X}function f() {byte.e(delete A.X);}}
```

```
contract D{constructor(){new D=D;}}
```

**Comment by Marenz** [comment]
At least the first two are kind of fixed by #9953 however, another assert is triggered. I am currently looking into this

**Comment by Marenz** [comment]
Okay, you actually removed the earlier one again. All that you listed here are fixed by #9953

**Comment by bshastry** [comment]
Sorry, mixed up and edited in place. Should you strike through the next time so the edit is seen.
